File: app/src/test/kotlin/tech/mmarca/openvitals/features/caffeine/CaffeineViewModelTest.kt
package tech.mmarca.openvitals.features.caffeine

import tech.mmarca.openvitals.data.repository.contract.FakePreferences
import io.mockk.coEvery
import io.mockk.coVerify
import io.mockk.mockk
import java.time.LocalDate
import java.time.ZoneId
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.delay
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import tech.mmarca.openvitals.core.period.DatePeriod
import tech.mmarca.openvitals.core.presentation.ScreenError
import tech.mmarca.openvitals.data.repository.contract.CaffeineRepository
import tech.mmarca.openvitals.data.repository.contract.NutritionRepository
import tech.mmarca.openvitals.domain.model.CaffeineEntry
import tech.mmarca.openvitals.domain.model.CaffeinePeriodData
import tech.mmarca.openvitals.domain.preferences.CaffeinePreferences
import tech.mmarca.openvitals.util.MainDispatcherRule

@OptIn(ExperimentalCoroutinesApi::class)
class CaffeineViewModelTest {

    @get:Rule
    val mainDispatcherRule = MainDispatcherRule()

    private val today = LocalDate.now()

    @Test
    fun `first load shows setup when caffeine exists and profile is incomplete`() = runTest {
        val vm = viewModel(
            repository = repo(entries = listOf(entryAt(today))),
            preferences = prefs(CaffeinePreferences(profileCompleted = false)),
            initialAnalyticsRange = CaffeineAnalyticsRange.TODAY,
        )

        assertFalse(vm.uiState.value.isLoading)
        assertTrue(vm.uiState.value.showSetup)
        assertEquals(100.0, vm.uiState.value.display.todayTotalMg, 0.001)
    }

    @Test
    fun `skipSetup stores completed defaults and hides setup`() = runTest {
        val preferences = prefs(CaffeinePreferences(profileCompleted = false))
        val vm = viewModel(
            repository = repo(entries = listOf(entryAt(today))),
            preferences = preferences,
            initialAnalyticsRange = CaffeineAnalyticsRange.TODAY,
        )

        vm.skipSetup()
        advanceUntilIdle()

        assertTrue(preferences.caffeinePreferences().profileCompleted)
        assertEquals(CaffeinePreferences.DefaultHalfLifeMinutes, preferences.caffeinePreferences().halfLifeMinutes)
        assertFalse(vm.uiState.value.showSetup)
    }

    @Test
    fun `preference updates rebuild display`() = runTest {
        val preferences = prefs(CaffeinePreferences(profileCompleted = true, sleepThresholdMg = 60))
        val vm = viewModel(
            repository = repo(entries = listOf(entryAt(today))),
            preferences = preferences,
            initialAnalyticsRange = CaffeineAnalyticsRange.TODAY,
        )

        preferences.setCaffeinePreferences(preferences.caffeinePreferences().copy(sleepThresholdMg = 35))
        advanceUntilIdle()

        assertEquals(35, vm.uiState.value.preferences.sleepThresholdMg)
        assertEquals(35, vm.uiState.value.display.sleepThresholdMg)
    }

    @Test
    fun `analytics range selection reloads matching caffeine window`() = runTest {
        val repository = repo()
        val vm = viewModel(
            repository = repository,
            preferences = prefs(CaffeinePreferences(profileCompleted = true)),
            initialAnalyticsRange = CaffeineAnalyticsRange.TODAY,
        )

        vm.selectAnalyticsRange(CaffeineAnalyticsRange.LAST_90_DAYS)

        assertEquals(CaffeineAnalyticsRange.LAST_90_DAYS, vm.uiState.value.analyticsRange)
        coVerify {
            repository.loadCaffeineData(
                DatePeriod(today.minusDays(89), today),
            )
        }

        // The same range again is not a reload.
        vm.selectAnalyticsRange(CaffeineAnalyticsRange.LAST_90_DAYS)
        advanceUntilIdle()

        coVerify(exactly = 1) {
            repository.loadCaffeineData(
                DatePeriod(today.minusDays(89), today),
            )
        }
    }

    @Test
    fun `refresh reloads`() = runTest {
        val repository = repo()
        val vm = viewModel(
            repository = repository,
            preferences = prefs(CaffeinePreferences(profileCompleted = true)),
        )

        vm.refresh()

        coVerify(atLeast = 2) { repository.loadCaffeineData(any()) }
    }

    @Test
    fun `newer load wins when analytics range requests overlap`() = runTest {
        val repository = mockk<CaffeineRepository>()
        coEvery { repository.loadCaffeineData(any()) } coAnswers {
            val period = firstArg<DatePeriod>()
            if (period.start == today.minusDays(89)) {
                delay(100)
            }
            CaffeinePeriodData(
                entries = listOf(entryAt(period.start, id = period.start.toString()))
            )
        }
        val vm = viewModel(
            repository = repository,
            preferences = prefs(CaffeinePreferences(profileCompleted = true)),
            initialAnalyticsRange = CaffeineAnalyticsRange.TODAY,
        )

        vm.selectAnalyticsRange(CaffeineAnalyticsRange.LAST_90_DAYS)
        vm.selectAnalyticsRange(CaffeineAnalyticsRange.TODAY)

        assertEquals(CaffeineAnalyticsRange.TODAY, vm.uiState.value.analyticsRange)
        assertEquals(today.toString(), vm.uiState.value.entries.single().id)
    }

    @Test
    fun `deleting a drink removes it optimistically and force-reloads`() = runTest {
        val entries = listOf(
            entryAt(today, id = "a", isOpenVitalsEntry = true),
            entryAt(today, id = "b", isOpenVitalsEntry = true),
        )
        val repository = mockk<CaffeineRepository>()
        // The reload returns the trimmed list, as Health Connect would after the delete.
        coEvery { repository.loadCaffeineData(any()) } returnsMany listOf(
            CaffeinePeriodData(entries),
            CaffeinePeriodData(entries.drop(1)),
        )
        val nutrition = mockk<NutritionRepository>()
        coEvery { nutrition.deleteNutritionEntry("a") } returns Unit

        val vm = viewModel(
            repository = repository,
            preferences = prefs(CaffeinePreferences(profileCompleted = true)),
            nutritionRepository = nutrition,
        )
        vm.deleteCaffeineEntry("a")

        assertEquals(listOf("b"), vm.uiState.value.entries.map { it.id })
        assertNull(vm.uiState.value.error)
        // A caffeine entry IS a nutrition record, so the nutrition repository is what deletes.
        coVerify { nutrition.deleteNutritionEntry("a") }
        coVerify(atLeast = 2) { repository.loadCaffeineData(any()) }
    }

    @Test
    fun `a failed delete restores the drink and surfaces the error`() = runTest {
        val entries = listOf(
            entryAt(today, id = "a", isOpenVitalsEntry = true),
            entryAt(today, id = "b", isOpenVitalsEntry = true),
        )
        val nutrition = mockk<NutritionRepository>()
        coEvery { nutrition.deleteNutritionEntry("a") } throws RuntimeException("denied")

        val vm = viewModel(
            repository = repo(entries = entries),
            preferences = prefs(CaffeinePreferences(profileCompleted = true)),
            nutritionRepository = nutrition,
        )
        vm.deleteCaffeineEntry("a")

        assertEquals(listOf("a", "b"), vm.uiState.value.entries.map { it.id })
        assertEquals(ScreenError.Message("denied"), vm.uiState.value.error)
    }

    @Test
    fun `a foreign or unidentified drink is never deleted`() = runTest {
        val nutrition = mockk<NutritionRepository>()
        val vm = viewModel(
            repository = repo(entries = listOf(entryAt(today, id = "foreign"))),
            preferences = prefs(CaffeinePreferences(profileCompleted = true)),
            nutritionRepository = nutrition,
        )

        vm.deleteCaffeineEntry("foreign")
        vm.deleteCaffeineEntry("")
        vm.deleteCaffeineEntry("missing")

        assertEquals(listOf("foreign"), vm.uiState.value.entries.map { it.id })
        coVerify(exactly = 0) { nutrition.deleteNutritionEntry(any()) }
    }

    @Test
    fun `a permission failure becomes ScreenError PermissionDenied`() = runTest {
        val repository = mockk<CaffeineRepository>()
        coEvery {
            repository.loadCaffeineData(any())
        } throws SecurityException("nutrition read")

        val vm = viewModel(
            repository = repository,
            preferences = prefs(CaffeinePreferences(profileCompleted = true)),
        )
        advanceUntilIdle()

        assertFalse(vm.uiState.value.isLoading)
        assertEquals(ScreenError.PermissionDenied, vm.uiState.value.error)
        assertTrue(vm.uiState.value.display.curvePoints.isEmpty())
    }

    @Test
    fun `an unexpected failure carries its message to the screen`() = runTest {
        val repository = mockk<CaffeineRepository>()
        coEvery {
            repository.loadCaffeineData(any())
        } throws RuntimeException("the provider hung up")

        val vm = viewModel(
            repository = repository,
            preferences = prefs(CaffeinePreferences(profileCompleted = true)),
        )
        advanceUntilIdle()

        assertFalse(vm.uiState.value.isLoading)
        assertEquals(ScreenError.Message("the provider hung up"), vm.uiState.value.error)
        // A failed load leaves the default insights in place, so there is nothing to draw.
        assertTrue(vm.uiState.value.display.curvePoints.isEmpty())
    }

    @Test
    fun `an empty load still gives the screen a display to render`() = runTest {
        val vm = viewModel(
            repository = repo(entries = emptyList()),
            preferences = prefs(CaffeinePreferences(profileCompleted = true)),
        )
        advanceUntilIdle()

        val state = vm.uiState.value
        assertFalse(state.isLoading)
        assertNull(state.error)
        assertEquals(CaffeineSleepImpactStatus.UNLIKELY, caffeineSleepImpactStatus(state.display))
        assertTrue(caffeineDistributionBars(state.analyticsDisplay.sourceTotals).isEmpty())
        // The curve is still plotted as a flat zero line, and the threshold line has to fit.
        assertTrue(state.display.curvePoints.isNotEmpty())
        assertTrue(
            caffeineCurveMaxMg(
                points = state.display.curvePoints,
                thresholdMg = state.display.sleepThresholdMg.toDouble(),
            ) > 0.0
        )
    }

    private fun viewModel(
        repository: CaffeineRepository,
        preferences: FakePreferences,
        initialAnalyticsRange: CaffeineAnalyticsRange = CaffeineAnalyticsRange.LAST_30_DAYS,
        nutritionRepository: NutritionRepository = mockk(),
    ): CaffeineViewModel =
        CaffeineViewModel(
            repository = repository,
            caffeineModel = preferences,
            bodyProfilePreferences = preferences,
            nutritionRepository = nutritionRepository,
            dispatchers = mainDispatcherRule.dispatcherProvider,
        ).also { vm ->
            // The screen opens on 30 days. A test that wants another range moves it, as the user would.
            if (initialAnalyticsRange != CaffeineAnalyticsRange.LAST_30_DAYS) vm.selectAnalyticsRange(initialAnalyticsRange)
        }

    private fun repo(entries: List<CaffeineEntry> = emptyList()): CaffeineRepository =
        mockk<CaffeineRepository>().also { repository ->
            coEvery { repository.loadCaffeineData(any()) } returns CaffeinePeriodData(entries)
        }

    private fun prefs(initial: CaffeinePreferences): FakePreferences =
        FakePreferences(initialCaffeine = initial)

    private fun entryAt(
        date: LocalDate,
        id: String = "coffee",
        caffeineMg: Double = 100.0,
        isOpenVitalsEntry: Boolean = false,
    ): CaffeineEntry {
        val start = date.atTime(8, 0).atZone(ZoneId.systemDefault()).toInstant()
        return CaffeineEntry(
            id = id,
            startTime = start,
            endTime = start.plusSeconds(10 * 60L),
            caffeineMg = caffeineMg,
            name = "Coffee",
            source = "test.source",
            mealType = 0,
            isOpenVitalsEntry = isOpenVitalsEntry,
        )
    }
}

File: test/features/caffeine/caffeine_view_model_test.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:shared_preferences/shared_preferences.dart';

import 'package:openvitals/core/period/period_load_query.dart';
import 'package:openvitals/core/period/time_range.dart';
import 'package:openvitals/core/presentation/screen_error.dart';
import 'package:openvitals/core/result/app_failure.dart';
import 'package:openvitals/core/result/result.dart';
import 'package:openvitals/data/prefs/preferences_repository.dart';
import 'package:openvitals/data/repository/contract/caffeine_repository.dart';
import 'package:openvitals/di/providers.dart';
import 'package:openvitals/domain/model/caffeine_models.dart';
import 'package:openvitals/domain/model/refresh_mode.dart';
import 'package:openvitals/features/caffeine/application/caffeine_display.dart';
import 'package:openvitals/features/caffeine/application/caffeine_view_model.dart';

/// Returns whatever it is told to, so the view-model's own behaviour — the
/// insight run, the display precompute, the failure mapping — is what is under
/// test.
class _FakeCaffeineRepository implements CaffeineRepository {
  _FakeCaffeineRepository(this.answer);

  Result<CaffeinePeriodData> answer;
  int loads = 0;
  RefreshMode? lastRefreshMode;
  DatePeriod? lastPeriod;

  @override
  Future<Result<CaffeinePeriodData>> loadCaffeineData(
    DatePeriod period, {
    RefreshMode refreshMode = RefreshMode.normal,
  }) async {
    loads += 1;
    lastRefreshMode = refreshMode;
    lastPeriod = period;
    return answer;
  }

  @override
  Future<Result<CaffeinePeriodData>> loadCaffeinePeriod(
    PeriodLoadQuery query, {
    RefreshMode refreshMode = RefreshMode.normal,
  }) async =>
      answer;

  @override
  dynamic noSuchMethod(Invocation invocation) => super.noSuchMethod(invocation);
}

CaffeineEntry _entry(DateTime start, double mg) => CaffeineEntry(
      id: 'entry-${start.millisecondsSinceEpoch}',
      startTime: start,
      endTime: start.add(const Duration(minutes: 10)),
      caffeineMg: mg,
      name: 'Coffee',
      source: 'Test source',
      mealType: 0,
    );

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  late _FakeCaffeineRepository repository;
  late ProviderContainer container;

  Future<ProviderContainer> boot(Result<CaffeinePeriodData> answer) async {
    SharedPreferences.setMockInitialValues(const <String, Object>{});
    final prefs = await SharedPreferences.getInstance();
    repository = _FakeCaffeineRepository(answer);
    container = ProviderContainer(overrides: [
      sharedPreferencesProvider.overrideWithValue(prefs),
      preferencesRepositoryProvider
          .overrideWithValue(PreferencesRepository(prefs)),
      caffeineRepositoryProvider.overrideWithValue(repository),
    ]);
    addTearDown(container.dispose);
    return container;
  }

  test('a loaded window lands with its display precomputed', () async {
    final now = DateTime.now();
    await boot(Ok(CaffeinePeriodData(entries: [
      _entry(now.subtract(const Duration(hours: 2)), 95),
      _entry(now.subtract(const Duration(hours: 6)), 120),
    ])));
    container.listen(caffeineProvider, (_, _) {});

    await container.read(caffeineProvider.notifier).load();
    // [build] self-triggers a first load; let it settle before asserting.
    await pumpEventQueue();

    final state = container.read(caffeineProvider);
    expect(state.isLoading, isFalse);
    expect(state.error, isNull);
    // The screen renders this; it must exist by the time loading ends.
    expect(state.display, isNotNull);
    // The curve is the whole point of the today card.
    expect(state.display!.home.insights.curvePoints, isNotEmpty);
    expect(state.display!.home.curveEntryTimes.length, 2);
    expect(state.display!.home.curveMaxMg, greaterThan(0));
    // Both drinks were today, so the analytics window sees them too.
    expect(state.display!.analytics.insights.periodTotalMg, 215.0);
    expect(state.display!.analytics.topSourceLabel, isNotNull);
    expect(state.display!.analytics.sourceBars.first.fraction, 1.0);
  });

  test('a permission failure becomes ScreenErrorPermissionDenied', () async {
    await boot(const Err(PermissionFailure('nutrition read')));
    container.listen(caffeineProvider, (_, _) {});

    await container.read(caffeineProvider.notifier).load();
    // [build] self-triggers a first load; let it settle before asserting.
    await pumpEventQueue();

    final state = container.read(caffeineProvider);
    expect(state.isLoading, isFalse);
    expect(state.error, const ScreenErrorPermissionDenied());
    expect(state.display, isNull);
  });

  test('an unexpected failure carries its message to the screen', () async {
    await boot(const Err(UnexpectedFailure('the provider hung up')));
    container.listen(caffeineProvider, (_, _) {});

    await container.read(caffeineProvider.notifier).load();
    // [build] self-triggers a first load; let it settle before asserting.
    await pumpEventQueue();

    expect(
      container.read(caffeineProvider).error,
      const ScreenErrorMessage('the provider hung up'),
    );
  });

  test('picking an analytics range reloads over the wider window', () async {
    await boot(const Ok(CaffeinePeriodData(entries: [])));
    container.listen(caffeineProvider, (_, _) {});
    final viewModel = container.read(caffeineProvider.notifier);
    await viewModel.load();
    await pumpEventQueue();

    final loadsBefore = repository.loads;
    viewModel.selectAnalyticsRange(CaffeineAnalyticsRange.last90Days);
    await pumpEventQueue();

    expect(container.read(caffeineProvider).analyticsRange,
        CaffeineAnalyticsRange.last90Days);
    expect(repository.loads, greaterThan(loadsBefore));
    expect(container.read(caffeineProvider).display, isNotNull);

    // The same range again is not a reload.
    final loadsAfter = repository.loads;
    viewModel.selectAnalyticsRange(CaffeineAnalyticsRange.last90Days);
    await pumpEventQueue();
    expect(repository.loads, loadsAfter);
  });

  test('refresh reloads in force mode', () async {
    await boot(const Ok(CaffeinePeriodData(entries: [])));
    container.listen(caffeineProvider, (_, _) {});
    final viewModel = container.read(caffeineProvider.notifier);

    await viewModel.load();
    await pumpEventQueue();
    await viewModel.refresh();

    expect(repository.lastRefreshMode, RefreshMode.force);
  });

  test('an empty load still gives the screen a display to render', () async {
    await boot(const Ok(CaffeinePeriodData(entries: [])));
    container.listen(caffeineProvider, (_, _) {});

    await container.read(caffeineProvider.notifier).load();
    // [build] self-triggers a first load; let it settle before asserting.
    await pumpEventQueue();

    final display = container.read(caffeineProvider).display!;
    expect(display.home.sleepImpactStatus, CaffeineSleepImpactStatus.unlikely);
    expect(display.analytics.sourceBars, isEmpty);
    expect(display.home.curveMaxMg, greaterThan(0));
  });
}
